### PR TITLE
Decimal quantization

### PR DESCRIPTION
**Pull request by kdeldycke**
This PR add a `decimal_quantization` parameter to `format_decimal()`, `format_currency()`, `format_percent()` and `format_scientific()`, to bypass the forced quantization on the fractional part. This PR definitely address #90 and is a continuation of #410.

This PR has been extensively reviewed and is waiting for #538 to be merged, so I can rebase that one on top of it.

### CODE REVIEW

**Comment by kdeldycke** [comment]
This PR was previously based on #491 and was partially duplicating it. Now that the later was merged, I rebased and squashed the whole code on top of the current `master` branch.

This PR is ready to be reviewed and merged.

**Comment by kdeldycke** [comment]
Thanks @benselme for the reviewing effort! I'll address all your comments in one or two weeks as I'm on holidays right now! 😎

File: babel/numbers.py [hunk 1]
```diff
@@ -304,15 +306,22 @@ def format_number(number, locale=LC_NUMERIC):
...
     """
-    # Do we really need this one?
+    warnings.warn('Use babel.numbers.format_decimal() instead.', DeprecationWarning)
     return format_decimal(number, locale=locale)
 
 
-def format_decimal(number, format=None, locale=LC_NUMERIC):
+def format_decimal(
+        number, format=None, locale=LC_NUMERIC,
+        decimal_quantization=True):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in d89365af925ce252b66af1f1510c5553baf29c87.

File: babel/numbers.py [hunk 2]
```diff
@@ -340,15 +349,17 @@ def format_decimal(number, format=None, locale=LC_NUMERIC):
...
 
 class UnknownCurrencyFormatError(KeyError):
     """Exception raised when an unknown currency format is requested."""
 
 
-def format_currency(number, currency, format=None, locale=LC_NUMERIC,
-                    currency_digits=True, format_type='standard'):
+def format_currency(
+        number, currency, format=None, locale=LC_NUMERIC, currency_digits=True,
+        format_type='standard', decimal_quantization=True):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in d89365af925ce252b66af1f1510c5553baf29c87.

File: babel/numbers.py [hunk 3]
```diff
@@ -412,17 +423,16 @@ def format_currency(number, currency, format=None, locale=LC_NUMERIC,
...
+                "%r is not a known currency format type" % format_type)
 
+    return pattern.apply(
+        number, locale, currency=currency, currency_digits=currency_digits,
+        decimal_quantization=decimal_quantization)
 
-def format_percent(number, format=None, locale=LC_NUMERIC):
+
+def format_percent(
+        number, format=None, locale=LC_NUMERIC, decimal_quantization=True):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in d89365af925ce252b66af1f1510c5553baf29c87.

File: babel/numbers.py [hunk 4]
```diff
@@ -445,10 +455,12 @@ def format_percent(number, format=None, locale=LC_NUMERIC):
...
         format = locale.percent_formats.get(format)
     pattern = parse_pattern(format)
-    return pattern.apply(number, locale)
+    return pattern.apply(
+        number, locale, decimal_quantization=decimal_quantization)
 
 
-def format_scientific(number, format=None, locale=LC_NUMERIC):
+def format_scientific(
+        number, format=None, locale=LC_NUMERIC, decimal_quantization=True):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in d89365af925ce252b66af1f1510c5553baf29c87.

File: babel/numbers.py [hunk 5]
```diff
@@ -641,68 +654,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
-        frac_prec = force_frac or self.frac_prec
+    @property
+    def scale(self):
+        """Scaling factor to apply to the number before rendering.
+
+        Auto-set to a factor of 2 or 3 if presence of a % or ‰ sign is
+        detected in the pattern. Default is to not mess with the scale at all.
+        """
+        scale = 0
+        if '%' in ''.join(self.prefix + self.suffix):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in 7e15be91aac9c9ef332e0431a2ef652970283915.

File: babel/numbers.py [hunk 6]
```diff
@@ -641,68 +654,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+            scale = 2
+        elif u'‰' in ''.join(self.prefix + self.suffix):
```
Comment (kdeldycke) [inline babel/numbers.py]: Done in 7e15be91aac9c9ef332e0431a2ef652970283915.

**Comment by kdeldycke** [comment]
I just addressed all coding and document issues. The PR has been squashed into one commit.

@benselme : I know this PR is not easy to review. To ease the pain I took the time to write expansive tests, searching for edge-cases and undefined situations.

I'm really happy with the results as I found a way to handle generically any rendering pattern.

Now the only remaining issue concerns the `#E0` situation, with breaks the consistency of the pattern rendering. What should I do here to have this PR accepted?

File: babel/numbers.py (hunk 5 above)
Comment (kdeldycke) [inline babel/numbers.py]: The call to `join` is indeed required as `prefix` and `suffix` are lists of strings. Removing the `join` introduce a regression. See: https://travis-ci.org/python-babel/babel/jobs/240689769#L1007

I reverted the change in 4887fdfa695a93b7d148057c167792f3a98a8912. All unittests passes now.

File: babel/numbers.py (hunk 6 above)
Comment (kdeldycke) [inline babel/numbers.py]: The call to `join` is indeed required as `prefix` and `suffix` are lists of strings. Removing the `join` introduce a regression. See: https://travis-ci.org/python-babel/babel/jobs/240689769#L1007

I reverted the change in 4887fdfa695a93b7d148057c167792f3a98a8912. All unittests passes now.

Comment (benselme) [inline babel/numbers.py]: Ok, sorry about this one !

Comment (kdeldycke) [inline babel/numbers.py]: No worries! :)

**Comment by benselme** [comment]
Thanks a lot for your work. 

What's here looks good to me. But I think the `#E0` issue needs to be solved before a merge can take place, since it makes scientific formatting useless for dozens of languages. I would write a special case for it, with proper comments and test cases. I think that's what the cldr elixir package [does](https://github.com/kipcole9/cldr/blob/master/lib/cldr/number/formatter/decimal_formatter.ex#L314), but I very well might be wrong. I had a quick look in ICU4J but could not locate the relevant code.

I would also love to have @etanol 's opinion on all this since he's the other person who recently did some work on numbers.

PS: the [CLDR-users mailing list](http://unicode.org/mailman/listinfo/cldr-users) is a very useful resource.

**Comment by etanol** [comment]
Sorry to respond so late, but I haven't been involved in Python development for a while.

The changes look good, although they are a bit difficult to review because refactoring and new functionality are **not** in separate commits.  If you have the time, please consider splitting this PR in to two commits, otherwise it's very confusing to tell apart code movements from actual new functionality or bugfixing.

I also have a couple of other comments for specific snippets in case you want to address them.

Overall, I'm okay with the feature as long as it doesn't break backwards compatibility.

File: babel/numbers.py [hunk 7]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
-        else:
-            self.scale = 0
 
     def __repr__(self):
         return '<%s %r>' % (type(self).__name__, self.pattern)
 
-    def apply(self, value, locale, currency=None, force_frac=None):
-        frac_prec = force_frac or self.frac_prec
+    @property
+    def scale(self):
```
Comment (etanol) [inline babel/numbers.py]: Turning a _precomputed_ field into a property is going to have a small performance impact.  My first motivation to contribute to this module was a use case we had, where currency formatting was consuming a lot of CPU time.  This is not Java, so we don't have a JIT that can inline this method call.

If you have time, try to do some trivial benchmarking by comparing against the parent commit.  Think about formatting hundreds of thousands of values.

File: babel/numbers.py [hunk 8]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+
+    def apply(
+            self, value, locale, currency=None, currency_digits=True,
+            decimal_quantization=True):
+        """Renders into a string a number following the defined pattern.
+
+        Forced decimal quantization is active by default so we'll produce a
+        number string that is strictly following CLDR pattern definitions.
+        """
+        # Manipulates decimal instances only.
```
Comment (etanol) [inline babel/numbers.py]: Most of the one-line comments in this method are not very useful.  I understand and appreciate the aesthetics, but in practice they don't provide any value.  I think it's better to add small paragraphs near the complex chunks of code, explaining the abstract ideas that lead to them.

File: babel/numbers.py [hunk 9]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+        # Force precision to the currency's.
+        if currency and currency_digits:
+            frac_prec = (get_currency_precision(currency), ) * 2
+
+        # Disabled decimal quantization will ensure no loss of precision when
+        # rendering the number.
+        if not decimal_quantization:
+            # Bump decimal precision to the natural precision of the number if
+            # it exceeds the one we're about to use.
+            frac_prec = (frac_prec[0], max([frac_prec[1], self.precision(value)]))
```
Comment (etanol) [inline babel/numbers.py]: So this is the core of the PR, right?  As I mentioned in the discussion, it would be much better to separate this (and all dependant changes) to a separate commit.  This will allow reviewers to choose granularity (since the GitHub UI allows to review all commits combined), and will ease potential bisections in the future.

File: babel/numbers.py (hunk 7 above)
Comment (akx) [inline babel/numbers.py]: This could use the [`@cached_property` decorator](https://github.com/bottlepy/bottle/commit/fa7733e075da0d790d809aa3d2f53071897e6f76#diff-ad8cb2f640fd3a70db3fc97f3044a4e6R147).

File: babel/numbers.py [hunk 10]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+        """
+        scale = 0
+        if '%' in ''.join(self.prefix + self.suffix):
+            scale = 2
+        elif u'‰' in ''.join(self.prefix + self.suffix):
+            scale = 3
+        return scale
+
+    @staticmethod
+    def precision(number):
```
Comment (akx) [inline babel/numbers.py]: I think method names should contain a verb – perhaps this should be `get_precision()`?

Also, instead of a static method, consider making it a free function?

File: babel/numbers.py [hunk 11]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+        Precision is extracted from the fractional part only.
+        """
+        # Copied from: https://github.com/mahmoud/boltons/pull/59
+        decimal_tuple = number.normalize().as_tuple()
+        if decimal_tuple.exponent >= 0:
+            return 0
+        return abs(decimal_tuple.exponent)
+
+    @staticmethod
+    def quantum(precision):
```
Comment (akx) [inline babel/numbers.py]: The comments for the `precision` method apply here too :)

File: tests/test_numbers.py [hunk 12]
```diff
@@ -16,7 +16,10 @@
 
 from datetime import date
 
-from babel import numbers
+from babel.numbers import (
```
Comment (akx) [inline tests/test_numbers.py]: Isn't this a repeat of line 23, below?

File: tests/test_numbers.py [hunk 13]
```diff
@@ -283,11 +292,51 @@ def test_format_decimal():
...
     assert numbers.format_decimal(12345.5, locale='en_US') == u'12,345.5'
+    assert numbers.format_decimal(0001.2345000, locale='en_US') == u'1.234'
+    assert numbers.format_decimal(-0001.2346000, locale='en_US') == u'-1.235'
+    assert numbers.format_decimal(0000000.5, locale='en_US') == u'0.5'
+    assert numbers.format_decimal(000, locale='en_US') == u'0'
+
+
+def test_format_decimal_quantization():
+    # Test precision conservation.
+    test_data = [
```
Comment (akx) [inline tests/test_numbers.py]: You can use [`@pytest.mark.parametrize`](https://docs.pytest.org/en/latest/parametrize.html) here:

```python
@pytest.mark.parametrize('input_value, expected', [
 ('10000', '10,000'),
 # (etc, etc.)
])
def test_format_decimal_quantization(input_value, expected):
  # ...
```

This has the advantage that each of these cases will show up (and be addressable) as a separate test case, so if one of those fails, the reader will not have to read the locals to figure out which particular case failed. :)

File: tests/test_numbers.py [hunk 14]
```diff
@@ -328,8 +383,43 @@ def test_format_currency_format_type():
             == u'1.099,98')
 
 
+def test_format_currency_quantization():
```
Comment (akx) [inline tests/test_numbers.py]: Same `parametrize` comment applies here.

File: tests/test_numbers.py [hunk 15]
```diff
@@ -339,14 +429,82 @@ def test_format_percent():
             == u'25,123\u2030')
 
 
-def test_scientific_exponent_displayed_as_integer():
-    assert numbers.format_scientific(100000, locale='en_US') == u'1E5'
+def test_format_percent_quantization():
```
Comment (akx) [inline tests/test_numbers.py]: And parametrize this too :)

File: tests/test_numbers.py [hunk 16]
```diff
@@ -339,14 +429,82 @@ def test_format_percent():
...
+    assert numbers.format_scientific(4234567, u'#.#E0', locale='en_US') == u'4.2E6'
+    assert numbers.format_scientific(4234567, u'0E0000', locale='en_US') == u'4E0006'
+    assert numbers.format_scientific(4234567, u'##0E00', locale='en_US') == u'4E06'
+    assert numbers.format_scientific(4234567, u'##00E00', locale='en_US') == u'42E05'
+    assert numbers.format_scientific(4234567, u'0,000E00', locale='en_US') == u'4,235E03'
+    assert numbers.format_scientific(4234567, u'##0.#####E00', locale='en_US') == u'4.23457E06'
+    assert numbers.format_scientific(42, u'00000.000000E0000', locale='en_US') == u'42000.000000E-0003'
+
+
+def test_format_scientific_quantization():
```
Comment (akx) [inline tests/test_numbers.py]: And this oughta get parametrized too :D

File: babel/numbers.py [hunk 17]
```diff
@@ -641,68 +698,134 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+        # Scale the number.
         value = value.scaleb(self.scale)
+
+        # Separate the absolute value from its sign.
         is_negative = int(value.is_signed())
-        if self.exp_prec:  # Scientific notation
+        value = abs(value).normalize()
+
+        # Prepare scientific notation metadata.
+        if self.exp_prec:
```
Comment (akx) [inline babel/numbers.py]: There's big chunks of scientific notation code interspersed with more "regular" rendering here – would you think the `exp_prec` bits could maybe be refactored into separate functions?

File: tests/test_numbers.py [hunk 18]
```diff
@@ -124,7 +127,7 @@ def test_scientific_notation(self):
         self.assertEqual(fmt, '1.2E3')
         # Exponent grouping
         fmt = numbers.format_scientific(12345, '##0.####E0', locale='en_US')
-        self.assertEqual(fmt, '12.345E3')
+        self.assertEqual(fmt, '1.2345E4')
```
Comment (akx) [inline tests/test_numbers.py]: So was this a bug, or a deliberate change in output?

File: babel/numbers.py (hunk 7 above)
Comment (kdeldycke) [inline babel/numbers.py]: I do not want to add other module dependencies so I abandoned the idea of reusing a `@cached_property` decorator.

Instead I simply kept the original method to keep the code readable and documented. See how the scaling value is now computed once since commit 7de43294594680763d329272c69e3509e5c41984.

File: babel/numbers.py (hunk 8 above)
Comment (kdeldycke) [inline babel/numbers.py]: I stand by the principle that you can't have too much documentation, but I don't really want to nit-pick here. Also note that this commenting style was originally there and I simply tried to keep the spirit of the current code base.

Anyway I made the necessary changes to remove one-line comments in commit 22b8f813166b062e012f1d878d8f11195ae483b7.

File: babel/numbers.py (hunk 10 above)
Comment (kdeldycke) [inline babel/numbers.py]: I renamed the method and made it a free function in commit 8c70916af3144e9d435a26736bea1f7c3efeb0a0.

File: babel/numbers.py (hunk 11 above)
Comment (kdeldycke) [inline babel/numbers.py]: Done in commit dba70a5f605256396146d218d471d81e553a8f1f.

File: tests/test_numbers.py (hunk 12 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: Good catch! Removed in commit 7a8c1f2dcd9007d9b348c7d1493a5f75fe947e0a.

File: babel/numbers.py (hunk 17 above)
Comment (kdeldycke) [inline babel/numbers.py]: Done in commit e66a4c1085eada2a68027ac95a5d8df4df05aed2.

File: tests/test_numbers.py (hunk 18 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: That is a deliberate change in output to fix what I think is a bug. The `##0.####E0` pattern implies only the first digit before the decimal separator is required.

File: tests/test_numbers.py (hunk 13 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: Done in commit b882589edd257a750c52e5015d742d41a70f2cf3.

File: tests/test_numbers.py (hunk 14 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: Done in commit b882589edd257a750c52e5015d742d41a70f2cf3.

File: tests/test_numbers.py (hunk 15 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: Done in commit b882589edd257a750c52e5015d742d41a70f2cf3.

File: tests/test_numbers.py (hunk 16 above)
Comment (kdeldycke) [inline tests/test_numbers.py]: Done in commit b882589edd257a750c52e5015d742d41a70f2cf3.

File: babel/numbers.py (hunk 9 above)
Comment (kdeldycke) [inline babel/numbers.py]: It is the core of the PR which is about bypassing decimal quantization. And I'll be happy to split it into another PR.

Because the code is currently highly intertwined, I'd rather complete the review process of the whole monolithic PR in its present form before attempting a code split.

**Comment by kdeldycke** [comment]
@benselme: I just addressed the issue of the default `#E0` rendering pattern in commit b27569040b67524af676cb71636668e9a2588754. The current PR is now rendering scientific notation like the CLDR elixir package and ICU4J does.

**Comment by kdeldycke** [comment]
I just finished addressing all pending comments of this PR.

What needs to be done now:
* Have @etanol, @akx and @benselme do a final pass on the PR and eventually lift their review process.
* Squash all commits into one.
* Split the PR into two: one for decimal quantization, and one for all the other refactoring and bug fixes.

File: babel/numbers.py [hunk 19]
```diff
@@ -641,68 +722,124 @@ def __init__(self, pattern, prefix, suffix, grouping,
...
+        frac_prec = self.frac_prec
+        if currency and currency_digits:
+            frac_prec = (get_currency_precision(currency), ) * 2
+
+        # Bump decimal precision to the natural precision of the number if it
+        # exceeds the one we're about to use. This adaptative precision is only
+        # triggered if the decimal quantization is disabled or if a scientific
+        # notation pattern has a missing mandatory fractional part (as in the
+        # default '#E0' pattern). This special case has been extensively
+        # disccused at https://github.com/python-babel/babel/pull/494#issuecomment-307649969 .
```
Comment (benselme) [inline babel/numbers.py]: Small typo here `discussed` not `disccused`

**Comment by benselme** [comment]
Thanks ! I'll try and have a look at your changes this week.

File: babel/numbers.py (hunk 19 above)
Comment (kdeldycke) [inline babel/numbers.py]: Good catch! Fixed in 6d0eabc18b176c8d8077c461f1f1b6272eca09a6.

**Comment by benselme** [comment]
Sorry, weeks go by and I've been quite busy. Not forgetting about this PR though. Thanks for your patience.

**Comment by kdeldycke** [comment]
Thanks @benselme for the feedback! I know this PR is kind of tough so take your time.

**Comment by kdeldycke** [comment]
Any news on that one? :)

File: babel/numbers.py (hunk 9 above)
Comment (etanol) [inline babel/numbers.py]: I think we are ready to see that split now.  And I'm going to insist more on separating refactoring from the real decimal quantization optional behavior code.  It will help potential future bisections.

Comment (kdeldycke) [reply babel/numbers.py]: Thanks @etanol and sorry for the delay.

I just spliced it up to #494.

**Comment by kdeldycke** [comment]
I just finished splitting this PR. #494 is now waiting to be merged so I can finally rebase that one to add proper decimal quantization control.

**Comment by kdeldycke** [comment]
This PR has been rebased on top of the latest `master` branch and is ready to be merged! :)

**Comment by etanol** [comment]
I think three years is enough proof of patience and perseverance.  Thanks @kdeldycke for having both.

**Comment by kdeldycke** [comment]
Thanks @etanol for the merge! 😃

**Comment by mishal193** [comment]
> This PR add a `decimal_quantization` parameter to `format_decimal()`, `format_currency()`, `format_percent()` and `format_scientific()`, to bypass the forced quantization on the fractional part. This PR definitely address #90 and is a continuation of #410.
> 
> 
> 
> This PR has been extensively reviewed and is waiting for #538 to be merged, so I can rebase that one on top of it.

(#538)
